lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp

```diff
@@ -0,0 +1,1084 @@
+    ++NumRetainsSunk;
+    recursivelyDeleteTriviallyDeadInstructions(R, true);
+  }
+  return Changed;
+}
+
+void RetainCodeMotionContext::mergeBBDataFlowStates(SILBasicBlock *BB) {
+  BlockState *State = BlockStates[BB];
+  State->BBSetIn.reset();
+  // If basic block has no predecessor, simply reset and return.
+  if (BB->pred_empty())
+    return;
+
+  // Intersect in all predecessors' BBSetOuts.
+  auto Iter = BB->pred_begin();
+  State->BBSetIn = BlockStates[*Iter]->BBSetOut;
+  Iter = std::next(Iter);
+  for (auto E = BB->pred_end(); Iter != E; ++Iter) {
+    State->BBSetIn &= BlockStates[*Iter]->BBSetOut;
+  }
+}
+
+bool RetainCodeMotionContext::processBBWithGenKillSet(SILBasicBlock *BB) {
+  RetainBlockState *State = BlockStates[BB];
+  // Compute the BBSetOut at the end of the basic block.
+  mergeBBDataFlowStates(BB);
+
+  // Compute the BBSetIn at the beginning of the basic block.
+  State->BBSetIn.reset(State->BBKillSet);
+  State->BBSetIn |= State->BBGenSet;
+ 
+  // If BBSetIn changes, then keep iterating until reached a fixed point.
+  return State->updateBBSetOut(State->BBSetIn);
+}
+
+void RetainCodeMotionContext::convergeCodeMotionDataFlow() {
+  // Process each basic block with the genset and killset. Every time the
+  // BBSetOut of a basic block changes, the optimization is rerun on its
+  // successors. 
+  llvm::SmallVector<SILBasicBlock *, 16> WorkList;
+  llvm::SmallPtrSet<SILBasicBlock *, 4> HandledBBs;
+  // Push into reverse post order so that we can pop from the back and get
+  // post order.
+  for (SILBasicBlock *B : PO->getReversePostOrder()) {
+    WorkList.push_back(B);
+    HandledBBs.insert(B);
+  }
+  while (!WorkList.empty()) {
+    SILBasicBlock *BB = WorkList.pop_back_val();
+    HandledBBs.erase(BB);
+    if (processBBWithGenKillSet(BB)) {
+      for (auto &X : BB->getSuccessors()) {
+        // We do not push basic block into the worklist if its already 
+        // in the worklist.
+        if (HandledBBs.count(X))
+          continue;
+        WorkList.push_back(X);
+      }
+    }
+  }
+}
+
+void RetainCodeMotionContext::computeCodeMotionInsertPoints() {
+  // The BBSetOuts have converged, run last iteration and figure out
+  // insertion point for each refcounted root.
+  for (SILBasicBlock *BB : PO->getReversePostOrder()) {
+    mergeBBDataFlowStates(BB);
+    RetainBlockState *S = BlockStates[BB];
+
+    // Compute insertion point generated by the edge value transition.
+    // If there is a transition from 1 to 0, that means we have a partial
+    // merge, which means the retain can NOT be sunk to the current block.
+    // place it at the predecessors.
+    for (unsigned i = 0; i < RCRootVault.size(); ++i) {
+      if (S->BBSetIn[i])
+        continue;
+      for (auto Pred : BB->getPreds()) {
+        BlockState *PBB = BlockStates[Pred];
+        if (!PBB->BBSetOut[i])
+          continue;
+        InsertPoints[RCRootVault[i]].push_back(Pred->getTerminator());
+      }
+    }
+
```

Comment:
If BBSetIn is all-zero, do we actually need to continue processing (and walking the entire block) or can we just exit?